src/core/layout/qgslayoutitemlegend.cpp

```diff
@@ -1133,14 +1133,29 @@ void QgsLayoutItemLegend::doUpdateFilterByMap()
 
     mapSettings.setExpressionContext( createExpressionContext() );
 
-    const QgsGeometry atlasGeometry = mInAtlas ? mLayout->reportContext().currentGeometry( mapSettings.destinationCrs() ) : QgsGeometry();
+    const QgsGeometry atlasGeometry { mLayout->reportContext().currentGeometry( mapSettings.destinationCrs() ) };
 
     QgsLayerTreeFilterSettings filterSettings( mapSettings );
 
+    QList<QgsMapLayer *> layersToClip;
+    if ( !atlasGeometry.isNull() && mMap->atlasClippingSettings()->enabled() )
+    {
+      layersToClip = mMap->atlasClippingSettings()->layersToClip();
+      for ( QgsMapLayer *layer : std::as_const( layersToClip ) )
+      {
+        QList<QgsMapLayer *> mapLayers { filterSettings.mapSettings().layers( true ) };
+        mapLayers.removeAll( layer );
+        filterSettings.mapSettings().setLayers( mapLayers );
```

Comment:
Why do we need that ? And why override it in mapsettings for each layer?

Reply:
It took me a while to get there too :)

Because of the convoluted logic inside the filter settings object which holds a mapsettings member object: it takes the layer from mapsettings with higher priority so we need to remove the layers we want to filter from the mapsettings if we want to use the overrided configuration from filtersettings.

Reply:
OK.  I get it!